### PR TITLE
ZK-4379: Provide a Loadingbar utility

### PR DESCRIPTION
**Pull request by plijyfes**

### CODE REVIEW

File: zk/src/org/zkoss/zk/ui/util/LoadingBar.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+import org.zkoss.zk.ui.sys.ExecutionCtrl;
+
+/**
+ * Utilities to send LoadingBar to the client.
+ *
+ * Note: this class makes accessing 'client-side' more directly.
+ *
+ * <h3>Custom Attributes</h3>
+ * <dl>
+ * <dt>org.zkoss.zk.ui.util.Toast.animationSpeed</dt>
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/util/LoadingBar.java]: Toast

File: zk/src/archive/web/js/zk/au.js [hunk 2]
```diff
@@ -1465,6 +1465,61 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
...
+	 * @param uuid the loadingbar uuid.
+	 * @param value the current value of the loadingbar.
+	 * @param position predefined positions.
+	 * @param indeterminate whether to start the indeterminate animation or not, default false.
+	 * @param animationSpeed animation speed for value changing in milliseconds, default 600ms.
+	 *
+	 * @since 9.0.0
+	 */
+	startLoadingBar: function (uuid, value, pos, indeterminate, animationSpeed) {
+		var loadingbar = (zul && zul.wgt) ? zul.wgt.Loadingbar : null;
```
Comment (plijyfes) [inline zk/src/archive/web/js/zk/au.js]: zk.load('zul.wgt',callback())

File: zk/src/archive/web/js/zk/au.js [hunk 3]
```diff
@@ -1465,6 +1465,61 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
...
+	 */
+	changeLoadingBar: function (uuid, value, indeterminate) {
+		var loadingbar = uuid ? zk.$('#' + uuid) : zk.$('@loadingbar');
+		if (loadingbar) {
+			if (value)
+				loadingbar.setValue(value);
+			if (indeterminate != null)
+				loadingbar.setIndeterminate(indeterminate);
+		} else {
+			zk.error('Target Loadingbar not exist');
```
Comment (plijyfes) [inline zk/src/archive/web/js/zk/au.js]: add uuid

File: zk/src/archive/web/js/zk/au.js [hunk 4]
```diff
@@ -1465,6 +1465,61 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
...
+	},
+	/**
+	 * Finish The loadingbar.
+	 * The loadingbar will be detached after the animation.
+	 *
+	 * @param uuid the target loadingbar uuid. If null, find the first loadingbar in the browser window.
+	 *
+	 * @since 9.0.0
+	 */
+	finishLoadingBar: function (uuid) {
```
Comment (plijyfes) [inline zk/src/archive/web/js/zk/au.js]: Loadingbar rename

File: zk/src/org/zkoss/zk/au/out/AuLoadingBar.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+import org.zkoss.zk.au.AuResponse;
+
+/**
+ * The au object for loadingbar.
+ *
+ * @author leon
+ * @since 9.0.0
+ */
+public class AuLoadingBar extends AuResponse {
```
Comment (plijyfes) [inline zk/src/org/zkoss/zk/au/out/AuLoadingBar.java]: rename

File: zk/src/org/zkoss/zk/ui/util/Clients.java [hunk 6]
```diff
@@ -684,4 +684,130 @@ public static final void showToast(String msg, String type, String position) {
 	public static final void showToast(String msg, String type, String position, int duration, boolean closable) {
 		Toast.show(msg, type, position, duration, closable);
 	}
+
```
Comment (plijyfes) [inline zk/src/org/zkoss/zk/ui/util/Clients.java]: remoce all

File: zk/src/org/zkoss/zk/ui/util/LoadingBar.java (hunk 1 above)
Comment (plijyfes) [inline zk/src/org/zkoss/zk/ui/util/LoadingBar.java]: doc

File: zk/src/org/zkoss/zk/ui/util/LoadingBar.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+		Executions.getCurrent().addAuResponse(response);
+	}
+
+	private static String getNextUuid() {
+		final Execution exec = Executions.getCurrent();
+		String uuid = "";
+		if (exec != null) {
+			ExecutionCtrl execCtrl = (ExecutionCtrl) exec;
+			DesktopCtrl desktopCtrl = (DesktopCtrl) exec.getDesktop();
+			uuid = desktopCtrl.getNextUuid(execCtrl.getCurrentPage());
```
Comment (plijyfes) [inline zk/src/org/zkoss/zk/ui/util/LoadingBar.java]: one line

File: zk/src/org/zkoss/zk/ui/util/LoadingBar.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+		}
+		return uuid;
+	}
+	
+	private static int getAnimationSpeed() {
+		if (_animationSpeed == null)
+			_animationSpeed = Library.getIntProperty("org.zkoss.zk.ui.util.LoadingBar.animationSpeed", 600);
```
Comment (plijyfes) [inline zk/src/org/zkoss/zk/ui/util/LoadingBar.java]: to ZKMAX & bar rename

File: zk/src/org/zkoss/zk/au/out/AuLoadingBar.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+	/**
+	 * Change the current value for the loadingbar.
+	 *
+	 * @param uuid the target loadingbar uuid. If null, find the first loadingbar in the browser window.
+	 * @param value the current value of the loadingbar.
+	 *
+	 * @since 9.0.0
+	 */
+	public AuLoadingBar(String uuid, int value) {
```
Comment (plijyfes) [inline zk/src/org/zkoss/zk/au/out/AuLoadingBar.java]: only one cunstruster: Auloadingbar(uuid, cmd ,.....)

File: zul/src/archive/web/js/zul/wgt/Loadingbar.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
+(() => {
+
+let _zkf: () => void;
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/Loadingbar.ts]: remove

File: zul/src/archive/web/js/zul/wgt/Loadingbar.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+		this._pos = opts.pos;
+		this._value = opts.value;
+		this._indeterminate = opts.indeterminate;
+		this._animationSpeed = opts.animationSpeed;
+	},
+	redraw: function (out) {
+		var uuid = this.uuid;
+		out.push('<div', this.domAttrs_(), '>');
+		out.push('<span id="', this.uuid, '-img"', 'class="', this.$s('image'),'"></span>');
+		out.push('<span class="', this.$s('effectbar'), ' ', this.$s('effectbar'),'"></span>'); // for indeterminate animation
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/Loadingbar.ts]: remove thrid bar

File: zul/src/archive/web/js/zul/wgt/Loadingbar.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+	$init: function (opts: {uuid: string; pos: string; value: number; indeterminate: boolean; animationSpeed: number}) {
+		this.$supers(zul.wgt.Loadingbar, '$init', arguments);
+		this.uuid = opts.uuid;
+		this._pos = opts.pos;
+		this._value = opts.value;
+		this._indeterminate = opts.indeterminate;
+		this._animationSpeed = opts.animationSpeed;
+	},
+	redraw: function (out) {
+		var uuid = this.uuid;
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/Loadingbar.ts]: not need

File: zul/src/archive/web/js/zul/wgt/Loadingbar.ts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+		out.push('<div', this.domAttrs_(), '>');
+		out.push('<span id="', this.uuid, '-img"', 'class="', this.$s('image'),'"></span>');
+		out.push('<span class="', this.$s('effectbar'), ' ', this.$s('effectbar'),'"></span>'); // for indeterminate animation
+		out.push('</div>');
+	},
+	finishCheck: function () {
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/Loadingbar.ts]: () {

File: zul/src/archive/web/js/zul/wgt/Loadingbar.ts [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+					width: this._value + '%'
+				}, { duration: this._animationSpeed, queue: false, easing: 'linear', done: self.finishCheck.bind(self) });
+			}
+		}
+	}
+}, {
+	/**
+	 * Start a LoadingBar.
+	 */
+	start(opts) {
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/Loadingbar.ts]: type check

File: zul/src/archive/web/js/zul/wgt/Loadingbar.ts [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+		if (!opts)
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/Loadingbar.ts]: if (opt) {....}
or if (!opt){ return; }

File: zul/src/archive/web/js/zul/wgt/Loadingbar.ts [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+			opts = {};
+		let loadingbar = new zul.wgt.Loadingbar(opts);
+		jq(this.getPositionWrapper_(opts.pos)).append(loadingbar);
+	},
+	getPositionWrapper_(pos: string): HTMLElement {
+		let cache: HTMLElement = this._positionWrappersCache[pos];
+		if (!cache) {
+			cache = jq(`<div class="z-loadingbar-wrapper z-loadingbar-wrapper-${pos}"></div>`)
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/Loadingbar.ts]: wrapper => top + bottom

File: zul/src/archive/web/js/zul/wgt/Loadingbar.ts [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+				.appendTo(document.body)
+				.get(0);
+			this._positionWrappersCache[pos] = cache;
+		}
+		return cache;
+	},
+	_positionWrappersCache: {}
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/Loadingbar.ts]: move to first line

File: zul/src/archive/web/js/zul/wgt/less/progressmeter.less [hunk 18]
```diff
@@ -35,3 +35,110 @@
 		}
 	}
 }
+
+// Loadingbar
+.z-loadingbar {
+	width: 100%;
+	height: 6px;
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/less/progressmeter.less]: "@LoadingbarHeight" in zkvar

File: zul/src/archive/web/js/zul/wgt/less/progressmeter.less [hunk 19]
```diff
@@ -35,3 +35,110 @@
...
+	background: transparent;
+	text-align: left;
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/less/progressmeter.less]: check default

File: zul/src/archive/web/js/zul/wgt/less/progressmeter.less [hunk 20]
```diff
@@ -35,3 +35,110 @@
...
+	overflow: hidden;
+	position: relative;
+	
+
+	&-image {
+		font-size: 0;
+		display: inline-block;
+		vertical-align: top;
+		height: 6px;
+		line-height: 0;
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/less/progressmeter.less]: check FF bug still exist

File: zul/src/archive/web/js/zul/wgt/less/progressmeter.less [hunk 21]
```diff
@@ -35,3 +35,110 @@
...
+		background-color: @loadingbarColor;
+	}
+
+	>.z-loadingbar-effectbar {
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/less/progressmeter.less]: &

File: zul/src/archive/web/js/zul/wgt/less/progressmeter.less [hunk 22]
```diff
@@ -35,3 +35,110 @@
...
+		display: none;
+	}
+
+	&-indeterminate {
+		opacity: 0.9;
+		width: 150%;
+
+		>.z-loadingbar-image {
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/less/progressmeter.less]: "> "

File: zul/src/archive/web/js/zul/wgt/less/progressmeter.less [hunk 23]
```diff
@@ -35,3 +35,110 @@
...
+			flex-direction: column-reverse;
+			top: unset;
+			bottom: 0;
+		}
+	}
+}
+
+.keyframes(@name,@rules) {
+
+	@-webkit-keyframes @name {
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/less/progressmeter.less]: check these need

File: zul/src/archive/web/js/zul/wgt/less/progressmeter.less [hunk 24]
```diff
@@ -35,3 +35,110 @@
...
+	}
+
+	&-wrapper {
+		display: -ms-flexbox;
+		display: flex;
+		position: fixed;
+		width: 100%;
+		left: 0;
+
+		>.z-loadingbar:nth-child(even) span {
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/wgt/less/progressmeter.less]: span -> class

File: zul/src/archive/web/zul/less/_zkvariables.less [hunk 25]
```diff
@@ -351,6 +351,11 @@
 @progressmeterBorderColor:     @colorGreyLight;
 @progressmeterBackgroundColor: @colorPrimary;
 
+// loadingbar
+@loadingbarColor:              @colorPrimary;
+@loadingbarSeparateColor:      @colorAccent;
```
Comment (scribetw) [inline zul/src/archive/web/zul/less/_zkvariables.less]: ???

**Pull request merged**